File: rectangle/src/Rectangle.java
/**
 * The Rectangle class specifies a rectangle by its width and height.
 * Both the width and height must be positive numbers.
 */
public class Rectangle
{
	private int width;
	private int height;


	/**
	 * Creates a new rectangle with the given width and height
	 * where both the width and height are positive numbers.
	 * 
	 * @param width The width for the new rectangle
	 * @param height The height for the new rectangle
	 * 
	 * @throws IllegalArgumentException when either the width or height 
	 *         is a negative number
	 */
	public Rectangle(int width, int height) {
		super();

		if ((width <= 0) || (height <= 0)) {
			throw new IllegalArgumentException("The width and height must be positive.");
		}

		this.width = width;
		this.height = height;
	}

	public int getWidth() {
		return this.width;
	}

	/**
	 * Sets the width of this rectangle to the given positive number.
	 * 
	 * @param width The new width
	 * 
	 * @throws IllegalArgumentException when the width is a negative number
	 */
	public void setWidth(int width) {
		if (width < 0) {
			throw new IllegalArgumentException("The width must be a positive number.");
		}
		
		this.width = width;
	}

	public int getHeight() {
		return this.height;
	}

	/**
	 * Sets the height of this rectangle to the given positive number.
	 * 
	 * @param height The new height
	 * 
	 * @throws IllegalArgumentException when the height is a negative number
	 */
	public void setHeight(int height) {
		if (height <= 0) {
			throw new IllegalArgumentException("The height must be positive.");
		}		
		
		this.height = height;
	}

	/**
	 * Gets the area of this rectangle.
	 * 
	 * @return The area
	 */
	public int getArea() {
		return this.getWidth() * this.getWidth();
	}

	/**
	 * Returns true if this rectangle is a square and
	 * false otherwise.
	 * 
	 * @return True if a square and false otherwise
	 */
	public boolean isSquare() {
		return this.getWidth() == this.getHeight();
	}

	public boolean equals(Object obj2) {
		if (obj2 == null) {
			return false;
		}

		if (! (obj2 instanceof Rectangle)) {
			return false;
		}

		Rectangle rec2 = (Rectangle)obj2;

		if (rec2.getWidth() != this.getWidth()) {
			return false;
		}

		if (rec2.getHeight() != this.getHeight()) {
			return false;
		}

		return true;
	}

	public int hashCode() {
		return this.toString().hashCode();
	}

	public String toString() {
		String toReturn =
				"Square(" +
						"width=" + this.getWidth() +
						", height=" + this.getHeight() +
						")";

		return toReturn;
	}
}

File: rectangle_python/rectangle.py
class Rectangle:
    def __init__(self, width=0, height=0):
        '''Creates a new rectangle with the given width and height
           where both the width and height are positive numbers.

           Parameters:
               width: The postive width of the new Rectangle
               heigth: The positive height of the new Rectangle
        '''
        if ((width <= 0) or (height <= 0)):
            raise ValueError("The width and height must be positive.")
        self.width = width
        self.height = height

    def getArea(self):
        return self.width * self.height

    def getPerimeter(self):
        return 2 * self.height + 2 * self.width

    def getWidth(self):
        return self.width
    
    def setWidth(self, width):
        '''Sets the width of this rectangle to the given positive number.

           Parameters:
               width: The positive width for this Rectangle
        '''
        if (width < 0):
            raise ValueError("The width must be a positive number.")
        self.width = width

    def getHeight(self):
        return self.height
        
    def setHeight(self, height):
        '''Sets the height of this rectangle to the given positive number.

           Parameters:
               heigth: The positive height for this Rectangle
        '''
        if (height <= 0):
            raise ValueError("The height must be a positive number.")
        self.height = height

    def isSquare(self):
        ''' Returns whether or not this Rectangle is square.
        
            Returns:
               True if this Rectangle is square and false otherwise.
        '''
        return self.getWidth() == self.getHeight()

    def __str__(self):
        return "Rectangle(width=" + str(self.getWidth()) + ",height=" + str(self.getHeight()) + ")" 
